In [122]:
import pandas as pd

<h2>Christian McCaffrey 2023

In [123]:
url_cmac = "https://www.pro-football-reference.com/players/M/McCaCh01.htm"
cmac = pd.read_html(url_cmac)
cmac = cmac[1]
display(cmac)

Unnamed: 0_level_0 Unnamed: 1_level_0 Unnamed: 2_level_0  \
                 Year                Age                 Tm   
0                2017                 21                CAR   
1                2018                 22                CAR   
2              2019*+                 23                CAR   
3                2020                 24                CAR   
4                2021                 25                CAR   
5               2022*                 26                2TM   
6                   *                NaN                CAR   
7                   *                NaN                SFO   
8                2023                 27                SFO   
9              Career             Career                NaN   
10              6 yrs              6 yrs                CAR   
11              2 yrs              2 yrs                SFO   

   Unnamed: 3_level_0 Unnamed: 4_level_0 Games     Rushing            ...  \
                  Pos                No.     G  GS     Att   Yds  TD  ...   
0                  RB               22.0    16  10     117   435   2  ...   
1                  RB               22.0    16  16     219  1098   7  ...   
2                  RB               22.0    16  16     287  1387  15  ...   
3                  RB               22.0     3   3      59   225   5  ...   
4                  RB               22.0     7   7      99   442   1  ...   
5                  RB                NaN    17  16     244  1139   8  ...   
6                  RB               22.0     6   6      85   393   2  ...   
7                  RB               23.0    11  10     159   746   6  ...   
8                  RB               23.0    14  14     244  1292  13  ...   
9                 NaN                NaN    89  82    1269  6018  51  ...   
10                NaN                NaN    64  58     866  3980  32  ...   
11                NaN                NaN    25  24     403  2038  19  ...   

   Receiving              Unnamed: 28_level_0 Unnamed: 29_level_0  \
         Y/G  Ctch% Y/Tgt               Touch               Y/Tch   
0       40.7  70.8%   5.8                 197                 5.5   
1       54.2  86.3%   7.0                 326                 6.0   
2       62.8  81.7%   7.1                 403                 5.9   
3       49.7  89.5%   7.8                  76                 4.9   
4       49.0  90.2%   8.4                 136                 5.8   
5       43.6  78.7%   6.9                 329                 5.7   
6       46.2  76.7%   6.4                 118                 5.7   
7       42.2  80.0%   7.1                 211                 5.7   
8       36.4  83.8%   7.5                 301                 6.0   
9       47.9  81.1%   6.9                1768                 5.8   
10      51.4  80.9%   6.8                1256                 5.8   
11      38.9  82.0%   7.3                 512                 5.9   

   Unnamed: 30_level_0 Unnamed: 31_level_0 Unnamed: 32_level_0  \
                  YScm                RRTD                 Fmb   
0                 1086                   7                   2   
1                 1965                  13                   4   
2                 2392                  19                   1   
3                  374                   6                   0   
4                  785                   2                   1   
5                 1880                  13                   1   
6                  670                   3                   0   
7                 1210                  10                   1   
8                 1801                  20                   3   
9                10283                  80                  12   
10                7272                  50                   8   
11                3011                  30                   4   

   Unnamed: 33_level_0       Unnamed: 34_level_0  
                    AV                    Awards  
0                  9.0                       Na

In [124]:
cmac.columns = cmac.columns.droplevel(0)
display(cmac)

Year     Age   Tm  Pos   No.   G  GS   Att   Yds  TD  ...   Y/G  Ctch%  \
0     2017      21  CAR   RB  22.0  16  10   117   435   2  ...  40.7  70.8%   
1     2018      22  CAR   RB  22.0  16  16   219  1098   7  ...  54.2  86.3%   
2   2019*+      23  CAR   RB  22.0  16  16   287  1387  15  ...  62.8  81.7%   
3     2020      24  CAR   RB  22.0   3   3    59   225   5  ...  49.7  89.5%   
4     2021      25  CAR   RB  22.0   7   7    99   442   1  ...  49.0  90.2%   
5    2022*      26  2TM   RB   NaN  17  16   244  1139   8  ...  43.6  78.7%   
6        *     NaN  CAR   RB  22.0   6   6    85   393   2  ...  46.2  76.7%   
7        *     NaN  SFO   RB  23.0  11  10   159   746   6  ...  42.2  80.0%   
8     2023      27  SFO   RB  23.0  14  14   244  1292  13  ...  36.4  83.8%   
9   Career  Career  NaN  NaN   NaN  89  82  1269  6018  51  ...  47.9  81.1%   
10   6 yrs   6 yrs  CAR  NaN   NaN  64  58   866  3980  32  ...  51.4  80.9%   
11   2 yrs   2 yrs  SFO  NaN   NaN  25  24   403  2038  19  ...  38.9  82.0%   

    Y/Tgt  Touch  Y/Tch   YScm  RRTD  Fmb    AV                    Awards  
0     5.8    197    5.5   1086     7    2   9.0                       NaN  
1     7.0    326    6.0   1965    13    4  16.0                       AP2  
2     7.1    403    5.9   2392    19    1  15.0        AP OPoY-3, AP1, PB  
3     7.8     76    4.9    374     6    0   3.0                       NaN  
4     8.4    136    5.8    785     2    1   5.0                       NaN  
5     6.9    329    5.7   1880    13    1  15.0  AP OPoY-7, AP CPoY-2, PB  
6     6.4    118    5.7    670     3    0   5.0  AP OPoY-7, AP CPoY-2, PB  
7     7.1    211    5.7   1210    10    1  10.0  AP OPoY-7, AP CPoY-2, PB  
8     7.5    301    6.0   1801    20    3   NaN                       NaN  
9     6.9   1768    5.8  10283    80   12  63.0                       NaN  
10    6.8   1256    5.8   7272    50    8  53.0                       NaN  
11    7.3    512    5.9   3011    30    4  10.0                       NaN  

[12 rows x 35 columns]

In [125]:
cmac_2023 = cmac.query('Year == "2023" ')
display(cmac_2023)

Year Age   Tm Pos   No.   G  GS  Att   Yds  TD  ...   Y/G  Ctch%  Y/Tgt  \
8  2023  27  SFO  RB  23.0  14  14  244  1292  13  ...  36.4  83.8%    7.5   

   Touch  Y/Tch  YScm  RRTD  Fmb  AV  Awards  
8    301    6.0  1801    20    3 NaN     NaN  

[1 rows x 35 columns]

In [126]:
cmac_2023 = cmac_2023[["Year", "G", "Yds", "TD"]]
display(cmac_2023)

Year   G   Yds  Yds  TD  TD
8  2023  14  1292  509  13   7

In [127]:
cmac_2023.columns = ["Year", "G", "Rush_Yds", "Rec_Yds", "Rush_TDs", "Rec_TDs"]
display(cmac_2023)

Year   G  Rush_Yds  Rec_Yds  Rush_TDs  Rec_TDs
8  2023  14      1292      509        13        7

In [128]:
cmac_2023.dtypes

Year        object
G            int64
Rush_Yds     int64
Rec_Yds      int64
Rush_TDs     int64
Rec_TDs      int64
dtype: object

In [129]:
cmac_2023["Total_Yds"] = cmac_2023["Rush_Yds"] + cmac_2023["Rec_Yds"]
cmac_2023["Total_TDs"] = cmac_2023["Rush_TDs"] + cmac_2023["Rec_TDs"]
cmac_2023["Player"] = "Christian McCaffrey"
display(cmac_2023)

Year   G  Rush_Yds  Rec_Yds  Rush_TDs  Rec_TDs  Total_Yds  Total_TDs  \
8  2023  14      1292      509        13        7       1801         20   

                Player  
8  Christian McCaffrey

In [130]:
cmac_2023_final = cmac_2023[["Year", "Player", "G", "Rush_Yds", "Rec_Yds", "Total_Yds", "Rush_TDs", "Rec_TDs",  "Total_TDs"]]
display(cmac_2023_final)

Year               Player   G  Rush_Yds  Rec_Yds  Total_Yds  Rush_TDs  \
8  2023  Christian McCaffrey  14      1292      509       1801        13   

   Rec_TDs  Total_TDs  
8        7         20

<h2>Adrian Peterson 2012

In [131]:
url_ad = "https://www.pro-football-reference.com/players/P/PeteAd01.htm"
ad = pd.read_html(url_ad)
ad = ad[0]
display(ad)

Unnamed: 0_level_0 Unnamed: 1_level_0 Unnamed: 2_level_0  \
                 Year                Age                 Tm   
0               2007*                 22                MIN   
1              2008*+                 23                MIN   
2              2009*+                 24                MIN   
3               2010*                 25                MIN   
4                2011                 26                MIN   
5              2012*+                 27                MIN   
6               2013*                 28                MIN   
7                2014                 29                MIN   
8              2015*+                 30                MIN   
9                2016                 31                MIN   
10               2017                 32                2TM   
11                NaN                NaN                ARI   
12                NaN                NaN                NOR   
13               2018                 33                WAS   
14               2019                 34                WAS   
15               2020                 35                DET   
16               2021                 36                2TM   
17                NaN                NaN                TEN   
18                NaN                NaN                SEA   
19             Career             Career                NaN   
20             10 yrs             10 yrs                MIN   
21              2 yrs              2 yrs                WAS   
22               1 yr               1 yr                ARI   
23               1 yr               1 yr                DET   
24               1 yr               1 yr                NOR   
25               1 yr               1 yr                SEA   
26               1 yr               1 yr                TEN   

   Unnamed: 3_level_0 Unnamed: 4_level_0 Games      Rushing              ...  \
                  Pos                No.     G   GS     Att    Yds   TD  ...   
0                  RB               28.0    14    9     238   1341   12  ...   
1                  RB               28.0    16   15     363   1760   10  ...   
2                  RB               28.0    16   15     314   1383   18  ...   
3                  RB               28.0    15   15     283   1298   12  ...   
4                  RB               28.0    12   12     208    970   12  ...   
5                  RB               28.0    16   16     348   2097   12  ...   
6                  RB               28.0    14   14     279   1266   10  ...   
7                  RB               28.0     1    1      21     75    0  ...   
8                  RB               28.0    16   16     327   1485   11  ...   
9                  RB               28.0     3    3      37     72    0  ...   
10                 RB                NaN    10    7     156    529    2  ...   
11                 RB               23.0     6    6     129    448    2  ...   
12                 RB               28.0     4    1      27     81    0  ...   
13                 RB               26.0    16   16     251   1042    7  ...   
14                 RB               26.0    15   15     211    898    5  ...   
15                 RB               28.0    16   10     156    604    7  ...   
16                 RB                NaN     4    3      38     98    2  ...   
17                 RB                8.0     3    2      27     82    1  ...   
18                 RB               21.0     1    1      11     16    1  ...   
19                NaN                NaN   184  167    3230  14918  120  ...   
20                NaN                NaN   123  116    2418  11747   97  ...   
21                NaN                NaN    31   31     462   1940   12  ...   
22                NaN                NaN     6    6     129    448    2  ...   
23                NaN                NaN    16   10     156    604    7  ...   
24                NaN                NaN     4    1      27     81    0  ...   
25             

In [132]:
ad.columns = ad.columns.droplevel(0)
print(ad)

      Year     Age   Tm  Pos   No.    G   GS   Att    Yds   TD  ...   Y/G  \
0    2007*      22  MIN   RB  28.0   14    9   238   1341   12  ...  19.1   
1   2008*+      23  MIN   RB  28.0   16   15   363   1760   10  ...   7.8   
2   2009*+      24  MIN   RB  28.0   16   15   314   1383   18  ...  27.3   
3    2010*      25  MIN   RB  28.0   15   15   283   1298   12  ...  22.7   
4     2011      26  MIN   RB  28.0   12   12   208    970   12  ...  11.6   
5   2012*+      27  MIN   RB  28.0   16   16   348   2097   12  ...  13.6   
6    2013*      28  MIN   RB  28.0   14   14   279   1266   10  ...  12.2   
7     2014      29  MIN   RB  28.0    1    1    21     75    0  ...  18.0   
8   2015*+      30  MIN   RB  28.0   16   16   327   1485   11  ...  13.9   
9     2016      31  MIN   RB  28.0    3    3    37     72    0  ...   2.7   
10    2017      32  2TM   RB   NaN   10    7   156    529    2  ...   7.0   
11     NaN     NaN  ARI   RB  23.0    6    6   129    448    2  ...  11.0   

In [133]:
ad_2012 = ad.query('Year == "2012*+" ')
display(ad_2012)

Year Age   Tm Pos   No.   G  GS  Att   Yds  TD  ...   Y/G  Ctch%  Y/Tgt  \
5  2012*+  27  MIN  RB  28.0  16  16  348  2097  12  ...  13.6  78.4%    4.3   

   Touch  Y/Tch  YScm  RRTD  Fmb  AV                                   Awards  
5    388    6.0  2314    13  4.0  19  AP MVP-1, AP OPoY-1, AP CPoY-2, AP1, PB  

[1 rows x 35 columns]

In [134]:
ad_2012 = ad_2012[["Year", "G", "Yds", "TD"]]
display(ad_2012)

Year   G   Yds    Yds  TD  TD
5  2012*+  16  2097  217.0  12   1

In [135]:
ad_2012.columns = ["Year", "G", "Rush_Yds", "Rec_Yds", "Rush_TDs", "Rec_TDs"]
display(ad_2012)

Year   G  Rush_Yds  Rec_Yds  Rush_TDs  Rec_TDs
5  2012*+  16      2097    217.0        12        1

In [136]:
ad_2012["Total_Yds"] = ad_2012["Rush_Yds"] + ad_2012["Rec_Yds"]
ad_2012["Total_TDs"] = ad_2012["Rush_TDs"] + ad_2012["Rec_TDs"]
ad_2012["Player"] = "Adrian Peterson" 
display(ad_2012)

Year   G  Rush_Yds  Rec_Yds  Rush_TDs  Rec_TDs  Total_Yds  Total_TDs  \
5  2012*+  16      2097    217.0        12        1     2314.0         13   

            Player  
5  Adrian Peterson

In [137]:
ad_2012_final = ad_2012[["Year", "Player", "G", "Rush_Yds", "Rec_Yds", "Total_Yds", "Rush_TDs", "Rec_TDs",  "Total_TDs"]]
display(ad_2012_final)

Year           Player   G  Rush_Yds  Rec_Yds  Total_Yds  Rush_TDs  \
5  2012*+  Adrian Peterson  16      2097    217.0     2314.0        12   

   Rec_TDs  Total_TDs  
5        1         13

In [138]:
cmac_ad = cmac_2023_final.append(ad_2012_final)
display(cmac_ad)

C:\Users\294747\AppData\Local\Temp\1\ipykernel_11940\2896180030.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  cmac_ad = cmac_2023_final.append(ad_2012_final)


Year               Player   G  Rush_Yds  Rec_Yds  Total_Yds  Rush_TDs  \
8    2023  Christian McCaffrey  14      1292    509.0     1801.0        13   
5  2012*+      Adrian Peterson  16      2097    217.0     2314.0        12   

   Rec_TDs  Total_TDs  
8        7         20  
5        1         13

<h2>With 1,801 yards currently, McCaffrey would have to average 171 total yards over the next 3 games to match Peterson's total in 2012.  Considering McCaffrey is averaging 129 total yards per game on the season, increasing that number by 42 yards for the remaining games would be tough task.  Another factor that's working against McCaffrey is that all of his receiving yards also count as passing yards for Purdy.  Given the pass-heavy era of the NFL that we are in, there will be some voters who view McCaffrey's receiving yards as a QB stat since Purdy is the one throwing him the ball.  Another thing of note: Peterson's 2,097 rushing yards in 2012 was just 8 yards away from Eric Dickerson's all-time record for most rushing yards in a season.  It feels like in today's era that a RB would need to have a special season in the rushing department like Peterson to have a solid shot at MVP. The gap between McCaffrey and Peterson in that department is pretty wide.